wptherml/exciton.py

```diff
@@ -0,0 +1,594 @@
+import numpy as np
+from matplotlib import pyplot as plt
+from .spectrum_driver import SpectrumDriver
+
+
+class ExcitonDriver(SpectrumDriver):
+    """A class for computing the dynamics and spectra of a system modelled by the Frenkel Exciton Hamiltonian
+
+    Attributes
+    ----------
+    radius : float
+        the radius of the sphere
+
+
+    Returns
+    -------
+    None
+
+    Examples
+    --------
+    >>> fill_in_with_actual_example!
+    """
+
+    def __init__(self, args):
+        self.parse_input(args)
+        # print("Exciton Energy is  ", self.exciton_energy)
+        # allocate the exciton Hamiltonian
+
+        self.exciton_hamiltonian = np.zeros(
+            (self.number_of_monomers, self.number_of_monomers)
+        )
+        self.c_vector = np.zeros(
+            (self.number_of_monomers, 1),  dtype=complex
+        )  # <== wavefunction coefficient vector
+        
+        # define cartesian coordinates of each monomer
+        self._compute_cartesian_coordinates()
+
+        # build exciton Hamiltonian
+        self.build_exciton_hamiltonian()
+
+        # Probably want to allow the user to specify an initial state!
+        # but right now just have the initial state with exciton localized on site 1
+        self.c_vector[0,0] = 1 + 0j
+        #self.density_matrix = np.dot(self.c_vector, np.conj(self.c_vector.T))
+        self.density_matrix = np.dot(self.c_vector, np.conj(self.c_vector.T))
+
+    def parse_input(self, args):
+        if "exciton_energy" in args:
+            self.exciton_energy = args["exciton_energy"]
+        else:
+            self.exciton_energy = 0.5
+
+        #############################################################
+        # going to change this to aggregate_shape 
+        #if "number_of_monomers" in args:
+        #    self.number_of_monomers = args["number_of_monomers"]
+        #else:
+        #    self.number_of_monomers = 2
+        #############################################################
+
+        if "aggregate_shape" in args:
+            self.aggregate_shape = args["aggregate_shape"]
+
+        # default shape is (2, 2, 1) -> 4 monomers total, 2 monomers displaced along x-axis and 2 along y-axis, all in plane (2D)
+        else:
+            self.aggregate_shape = (2, 2, 1)
+        
+        # user might accidentally make one element of the tuple zero
+        if self.aggregate_shape[0] == 0 or self.aggregate_shape[1] == 0 or self.aggregate_shape[2] == 0:
+            print(" Invalid shape!  All directions must have at least 1 layer!")
+            print(" Smallest possible shape (1,1,1) is a monomer - please check your input and try again")
+            exit
+
+        self.number_of_monomers = self.aggregate_shape[0] * self.aggregate_shape[1] * self.aggregate_shape[2]
+        # allow the user to specify the displacements along each axis (x, y, z) as a vector
+        if "displacement_vector" in args:
```

Comment:
I also included the horizontal and vertical displacements into this displacement vector:
horizontal_displacement <-> displacement_vector[0]
vertical_displacement <-> displacement_vector[1]

and if one wants 3D aggreagates (corresponding to some aggregate_shape (Nx, Ny, Nz) with Nz >= 2)
then the displacement along the z axis is given by displacement_vector[2]